functions/access.js

```diff
@@ -0,0 +1,136 @@
+        // Record that the token was checked
+        if (!granted.tokens[token].usedAt) {
+          granted.tokens[token].usedAt = [];
+        }
+        granted.tokens[token].usedAt.push(Date.now());
+        transaction.update(accessRef, granted);
+
+        return true;
+      }
+
+      return false;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
+      return false;
+    });
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  if (!venueId || !uid || !password || !email || !code) return {};
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+
+  const token = uuid();
+  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
+
+  const tokenData = {
+    usedAt: [Date.now()],
+    password: password || null,
+    email: email || null,
+    code: code || null,
+  };
+
+  if (grantedDoc.exists) {
+    await grantedDoc.update({ [`tokens.${token}`]: tokenData });
+  } else {
+    const granted = { tokens: { [token]: tokenData } };
+    await grantedDoc.set(granted);
+  }
+  return { token };
+};
+
+exports.checkAccess = functions.https.onCall(async (data, context) => {
+  console.log("checkAccess", data, context.auth);
+  const isValidToken =
+    context &&
+    context.auth &&
+    context.auth.uid &&
+    (await checkIsValidToken(data.venueId, context.auth.uid, data.token));
+  if (isValidToken) {
+    console.log(`valid token, returning ${data.token}`);
+    return { token: data.token };
+  }
+
+  const isPasswordValid =
+    data.password && (await isValidPassword(data.venueId, data.password));
+  const isEmailValid =
+    data.email && (await isValidEmail(data.venueId, data.email));
+  const isCodeValid = data.code && (await isValidCode(data.venueId, data.code));
```

Comment:
Along with the changes suggested to make `isValidPassword`/etc more robust (see other comments in this review)

We can make these all run in parallel if we use `Promise.all`:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

```suggestion
  const [isPasswordValid, isEmailValid, isCodeValid] = Promise.all([
    isValidPassword(data.venueId, data.password),
    isValidEmail(data.venueId, data.email),
    isValidCode(data.venueId, data.code),
  ]);
```

Reply:
Since each of these helper functions is individually calling `getAccessDoc`, which will then retrieve `venue` and check if `venue.exists`, it might be more efficient for us to refactor that check out of `getAccessDoc`, so that we can explicitly make it only once, rather than 3 times.

I haven't thoroughly looked through the code, but I think both `getAccessDoc` and each of the `isValidPassword`/etc helpers are only called from this function? If so.. I wonder if we can just collapse them all back down into a simpler location/check (or maybe even just inline in this function definition if that still leaves this code readable?)